### PR TITLE
[quant][graph] Add quant fusion for dynamic quantization

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35586 [quant][graph] Add quant fusion for dynamic quantization**
* #35448 [quant][graph] Add pass to insert quant dequant for dynamic quantization

Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20755680](https://our.internmc.facebook.com/intern/diff/D20755680)

### CODE REVIEW

File: test/test_quantize_script.py [hunk 1]
```diff
@@ -168,5 +170,21 @@ def forward(self, x):
                    .check("return") \
                    .run(str(get_forward_graph(m.fc1._c)))
 
+    def test_finalize_for_linear_dynamic(self):
```
Comment (jerryzh168) [inline test/test_quantize_script.py]: should be: dynamic_linear

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -2485,7 +2485,15 @@ void SwapDeQuant(std::shared_ptr<Graph>& graph) {
...
-void QuantFusion(std::shared_ptr<Graph>& graph) {
+void QuantFusion(std::shared_ptr<Graph>& graph, bool is_dynamic) {
+  if (is_dynamic) {
+    for (const auto& item : dynamic_quant_fusion_pattern_and_replacements()) {
+      SubgraphRewriter rewriter;
+      rewriter.RegisterRewritePattern(item.first, item.second);
+      rewriter.runOnGraph(graph);
+    }
+    return;
+  }
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp]: could you use the same code?
Also I think you'll need to rebase, return type for quant_fusion_pattern_and_replacements is changed

File: torch/quantization/_quantize_script.py [hunk 3]
```diff
@@ -115,3 +123,16 @@ def quantize_script(model, qconfig_dict, run_fn, run_args, inplace=False, debug=
     run_fn(model._c._get_method('forward'), *run_args)
     model = convert_script(model, True, debug)
     return model
+
+def quantize_dynamic_script(model, qconfig_dict, run_fn, run_args, debug=False):
```
Comment (jerryzh168) [inline torch/quantization/_quantize_script.py]: could you reuse the `quantize_script` code?

File: torch/quantization/_quantize_script.py [hunk 4]
```diff
@@ -95,6 +95,14 @@ def convert_script(model, inplace=False, debug=False):
         model = wrap_cpp_module(torch._C._jit_pass_quant_finalize(model._c))
     return model
 
+def convert_dynamic_script(model, debug=False):
```
Comment (jerryzh168) [inline torch/quantization/_quantize_script.py]: same here, could you reuse the code?

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
I think we should reuse most of the code

File: torch/quantization/_quantize_script.py [hunk 5]
```diff
@@ -95,26 +99,44 @@ def prepare_dynamic_script(model, qconfig_dict):
...
+def _quantize_script(model, qconfig_dict, run_fn, run_args, is_dynamic, debug):
+    _check_is_script_module(model)
+    _check_forward_method(model)
     torch._C._jit_pass_dedup_module_uses(model._c)
     model = wrap_cpp_module(torch._C._jit_pass_fold_convbn(model._c))
-    model = prepare_script(model, qconfig_dict, True)
+    if is_dynamic:
+        model = prepare_dynamic_script(model, qconfig_dict)
+    else:
+       model = prepare_script(model, qconfig_dict, True)
```
Comment (jerryzh168) [inline torch/quantization/_quantize_script.py]: identation

File: torch/quantization/_quantize_script.py [hunk 6]
```diff
@@ -95,26 +99,44 @@ def prepare_dynamic_script(model, qconfig_dict):
                                                                 True))
     return model
 
-def convert_script(model, inplace=False, debug=False):
+def _convert_script(model, is_dynamic, debug=False):
```
Comment (jerryzh168) [inline torch/quantization/_quantize_script.py]: could you refactor prepare as well?

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
please fix indentations, there are a few places were indentations are wrong

File: torch/quantization/_quantize_script.py [hunk 7]
```diff
@@ -74,47 +78,61 @@ def script_qconfig(qconfig):
...
+        model = prepare_dynamic_script(model, qconfig_dict)
+    else:
+        model = prepare_script(model, qconfig_dict, True)
     run_fn(model._c._get_method('forward'), *run_args)
-    model = convert_script(model, True, debug)
+
+    if is_dynamic:
+        model = convert_dynamic_script(model, debug)
+    else:
+        model = convert_script(model, True, debug)
```
Comment (jerryzh168) [inline torch/quantization/_quantize_script.py]: could you merge the two ifs?

File: test/test_quantization.py [hunk 8]
```diff
@@ -1081,6 +1081,43 @@ def test_skip_quant(self):
...
+            model_quantized = quantize_dynamic_script(
+                model_under_test,
+                qconfig_dict,
+                test_only_eval_fn,
+                [self.calib_data])
+            self.assertEqual(model_quantized(self.calib_data[0][0]), result_eager)
+
+            # Check to make sure choose_qparams->quant->dequant->linear is numerically
+            # equivalent to the final quantized model.
+            model_fake_quantized = quantize_dynamic_script(
```
Comment (jerryzh168) [inline test/test_quantization.py]: you mean model_dynamic_quantized?

Comment (jerryzh168) [reply test/test_quantization.py]: Oh, this is debug, could you use another name? fake_quantize has a special meaning in quantization

**Review by jerryzh168 — APPROVED** [review]
LG, some comments are inline

File: test/test_quantization.py (hunk 8 above)
Comment (supriyar) [reply test/test_quantization.py]: No, this tests the fake quantized model with the `quantize_per_tensor` op followed by float Linear op to be numerically equivalent to the dynamically quantized op in eager mode.

**Pull request closed without merging**